src/Samples/SubModelSeq/Program.fs

```diff
@@ -13,177 +29,192 @@ module List =
         | a when a = j -> i
         | a -> a)
 
+  let swapWithNext i = swap i (i + 1)
+  let swapWithPrev i = swap i (i - 1)
 
+  let cons a ma = a :: ma
+
+        
 [<AutoOpen>]
-module Domain =
+module Identifiable =
+
+  type Identifiable<'a> =
+    { Id: Guid
+      Value: 'a }
+  
+  module Identifiable =
 
-  type CounterId = CounterId of Guid
+    let getId m = m.Id
+    let get m = m.Value
+    let set v m = { m with Value = v }
+    let map f = f |> map get set
+    let mapIf p f = f |> map |> mapIf p
+    let mapIfId id f = f |> mapIf (fun m -> m.Id = id)
+
+
+[<AutoOpen>]
+module Counter =
 
   type Counter =
-    { Id: CounterId
-      CounterValue: int
+    { Count: int
       StepSize: int }
 
+  type CounterMsg =
+    | Increment
+    | Decrement
+    | SetStepSize of int
+    | Reset
+
   module Counter =
 
-    let create () =
-      { Id = Guid.NewGuid () |> CounterId
-        CounterValue = 0
+    let init =
+      { Count = 0
         StepSize = 1 }
+    
+    let canReset = (<>) init
+    
+    let update msg m =
+      match msg with
+      | Increment -> { m with Count = m.Count + m.StepSize }
+      | Decrement -> { m with Count = m.Count - m.StepSize }
+      | SetStepSize x -> { m with StepSize = x }
+      | Reset -> init
 
-    let increment c =
-      { c with CounterValue = c.CounterValue + c.StepSize }
 
-    let decrement c =
-      { c with CounterValue = c.CounterValue - c.StepSize}
 
-    let setStepSize step c =
-      { c with StepSize = step }
+[<AutoOpen>]
+module RoseTree =
 
-    let canReset c =
-      c.CounterValue <> 0 || c.StepSize <> 1
+  type RoseTree<'a> =
+    { Data: 'a
+      Children: RoseTree<'a> list }
 
-    let reset c =
-      { create () with Id = c.Id }
 
+  module RoseTree =
 
-module Tree =
+    let getData m = m.Data
+    let setData v m = { Data = v; Children = m.Children }
+    let rec mapData f n =
+      { Data = n.Data |> f
+        Children = n.Children |> (f |> mapData |> List.map) }
 
-  type Node<'a> =
-    { Data: 'a
-      Children: Node<'a> list }
+    let getChildren m = m.Children
+    let setChildren v m = { m with Children = v }
+    let mapChildren f = map getChildren setChildren f
 
-  let asLeaf a =
-    { Data = a
-      Children = [] }
+    let rec map f t =
+      t |> f |> mapChildren (f |> map |> List.map)
 
-  let dataOfChildren n =
-    n.Children |> List.map (fun nn -> nn.Data)
+    let asLeaf a =
+      { Data = a
+        Children = [] }
 
-  let rec map f n =
-    let nn = n |> f
-    { nn with Children = nn.Children |> List.map (map f) }
+    let addChild a = a |> asLeaf |> List.cons |> mapChildren
+    
+    let dataOfChildren n =
+      n.Children |> List.map getData
+    
+    let rec preorderFlatten n =
+      n :: List.collect preorderFlatten n.Children
```

Comment:
Why `preorder`? I don't understand the name. Is this depth-first? Could `depthFirstFlatten` be a better name?

Reply:
Yes, this is a depth-first search.  There are at least [six orderings](https://en.wikipedia.org/wiki/Depth-first_search#Vertex_orderings) when doing a DFS, and preorder is one of them.  In that sense, the name is helpful to understand the implementation.

On the other hand, I just checked, and as one might guess, the two call sites in this sample don't depend on the order (...so any DFS or any BFS would do).  For the sake of the sample then, we could simplify the name further to just `flatten`.

Want to simply call it `flatten`?

Reply:
Yes, personally I would prefer that since it's irrelevant.

Reply:
Done